avail-server/src/main/kotlin/com/avail/server/io/files/ServerFileWrapper.kt

```diff
@@ -0,0 +1,409 @@
+			if (file.isOpen)
+			{
+				try
+				{
+					file.close()
+				}
+				catch (e: IOException)
+				{
+					// Do nothing
+				}
+			}
+			if (!Files.deleteIfExists(Paths.get(path)))
+			{
+				failure(ServerErrorCode.FILE_NOT_FOUND, null)
+			}
+		}
+		catch (e: IOException)
+		{
+			failure(ServerErrorCode.IO_EXCEPTION, e)
+		}
+	}
+
+	init
+	{
+		FileManager.executeFileTask (Runnable {
+			val p = Paths.get(path)
+			val mimeType = Tika().detect(p)
+			file = createFile(path, fileChannel, mimeType, this)
+		})
+	}
+
+	/** Close this [ServerFileWrapper]. */
+	fun close () { file.close() }
+
+	/**
+	 * Provide the [raw bytes][AvailServerFile.rawContent] of the enclosed
+	 * [AvailServerFile] to the requesting consumer.
+	 *
+	 * @param id
+	 *   The [FileManager.fileCache] [UUID] that uniquely identifies the file.
+	 * @param consumer
+	 *   A function that accepts the [FileManager.fileCache] [UUID] that
+	 *   uniquely identifies the file and the
+	 *   [raw bytes][AvailServerFile.rawContent] of an [AvailServerFile].
+	 * @param failureHandler
+	 *   A function that accepts a [ServerErrorCode] that describes the nature
+	 *   of the failure and an optional [Throwable]. TODO refine error handling
+	 */
+	fun provide(
+		id: UUID,
+		consumer: (UUID, ByteArray) -> Unit,
+		failureHandler: (ServerErrorCode, Throwable?) -> Unit)
+	{
+		// TODO should be run on a separate thread or is that handled inside the
+		//  consumer?
+		if (!isLoadingFile)
+		{
+			// Opportunistic
+			consumer(id, file.rawContent)
+			return
+		}
+		synchronized(this)
+		{
+			if (!isLoadingFile)
+			{
+				// Opportunistic
+				consumer(id, file.rawContent)
+				return
+			}
+			handlerQueue.add(
+				FileRequestHandler(id, consumer, failureHandler))
+		}
+	}
+
+	/**
+	 * Notify this [ServerFileWrapper] that the [AvailServerFile] has been fully
+	 * read.
+	 */
+	fun notifyReady ()
+	{
+		synchronized(this)
+		{
+			isLoadingFile = false
+		}
+		// TODO should each be run on a separate thread or is that handled
+		//  inside the consumer?
+		val fileBytes = file.rawContent
+		handlerQueue.forEach { it.requestConsumer(it.id, fileBytes) }
+	}
+
+	/**
+	 * Notify this [ServerFileWrapper]
+	 */
+	fun notifyFailure ()
+	{
+		// TODO figure out input and how to handle this
+		//  this includes handling new requests...
+	}
+
+	companion object
+	{
+		/**
+		 * Answer a new [AvailServerFile] that will be associated with the
+		 * [ServerFileWrapper].
+		 *
+		 * @param path
+		 *   The String path to the target file on disk.
+		 * @param file
+		 *   The [AsynchronousFileChannel] used to access the file.
+		 * @param mimeType
+		 *   The MIME type of the file.
+		 * @param serverFileWrapper
+		 *   The owning `ServerFileWrapper`.
+		 */
+		fun createFile (
+			path: String,
+			file: AsynchronousFileChannel,
+			mimeType: String,
+			serverFileWrapper: ServerFileWrapper): AvailServerFile
+		{
+			return when (mimeType)
+			{
+				"text/plain", "text/json" ->
```

Comment:
Note that it's especially important to handle `"text/avail"`, which we'll need to augment Tika to recognize. I suppose that they provide an extension mechanism to make this is possible somehow? We will just want to use the `.avail` extension and the presence of UTF-8 content as the trigger for `"text/avail"`.

Reply:
this has been added in the resource folder of the avail server module.